### PR TITLE
Enforce metadata[fqdn] and ID presence in message

### PR DESCRIPTION
**Pull request by dLobatog**
https://projects.engineering.redhat.com/browse/RHICOMPL-499

### CODE REVIEW

File: app/services/xccdf_report_parser.rb [hunk 1]
```diff
@@ -30,6 +30,11 @@ def check_report_format
     )
   end
 
+  def valid_message_format?(message)
+    message['id'].present? && message['metadata'].present? &&
+      message['metadata']['fqdn'].present?
```
Comment (akofink) [inline app/services/xccdf_report_parser.rb]: This is a good case for `dig`:

```suggestion
    message['id'].present? && message.dig('metadata', 'fqdn').present?
```

File: app/services/host_inventory_api.rb [hunk 2]
```diff
@@ -24,6 +24,14 @@ def host_already_in_inventory(hostname_or_id)
     find_results(JSON.parse(response.body))
   end
 
+  def inventory_host
+    @inventory_host ||= host_already_in_inventory(@id) ||
+                        (@hostname && host_already_in_inventory(@hostname))
```
Comment (akofink) [inline app/services/host_inventory_api.rb]: We can/should remove this find by hostname. If we don't have ID, we fail to parse.

File: app/services/host_inventory_api.rb [hunk 3]
```diff
@@ -24,6 +24,14 @@ def host_already_in_inventory(hostname_or_id)
...
+  end
+
+  # This function is meant to be only used for testing purposes within a
+  # development environment. It is not and should not be called from anywhere
+  # else than a local rake task or the Rails console itself.
   def create_host_in_inventory
```
Comment (akofink) [inline app/services/host_inventory_api.rb]: I use httpie for this. I want to push for just removing it, but I can't really make a good case for why leaving this here is bad, other than it's no longer used. :man_shrugging: 

```
http --verify no --auth <auth> POST https://ci.cloud.redhat.com/api/inventory/v1/hosts < systems.json
HTTP/1.1 207 MULTI STATUS
Transfer-Encoding: chunked
X-Powered-By: Express
connection: close
content-type: application/json
date: Wed, 11 Mar 2020 16:15:39 GMT
server: openresty/1.13.6.1
set-cookie: bd2e829019cc5851647126f4f0834941=546c250d4ef07df905cfb33195ad89ad; path=/; HttpOnly; Secure
x-content-type-options: nosniff
x-fakamai-app-hint: API
x-fakamai-downstream-host: https://api-gateway-ci.5a9f.insights-dev.openshiftapps.com
x-fakamai-env: ci
x-rh-insights-request-id: 4400338bbd814d0f88a73882ba2894f8

{
    "data": [
        {
            "host": {
                ...
            },
            "status": 200
        }
    ],
    "errors": 0,
    "total": 1
}
```

With systems.json like this:

```js
[
  {
    "account": "1212729",
    "display_name": "rhel7-insights-client-fake100.virbr0.akofink-laptop",
    "facts": [],
    "fqdn": "rhel7-insights-client-fake001.virbr0.akofink-laptop",
    "reporter": "akofink's local dev machine",
    "stale_timestamp": "2021-03-04T16:15:44Z"
  }
]
```

File: app/services/concerns/xccdf/hosts.rb [hunk 4]
```diff
@@ -48,7 +48,6 @@ def test_result_profile
     def inventory_host
       @inventory_host ||= ::HostInventoryAPI.new(
         @host_inventory_id,
-        report_host,
```
Comment (akofink) [inline app/services/concerns/xccdf/hosts.rb]: This was the only usage of this method in our code - we should delete the method as well.

File: app/services/host_inventory_api.rb [hunk 5]
```diff
@@ -6,59 +6,32 @@
...
     @account = account
     @b64_identity = b64_identity || @account.b64_identity
   end
 
-  def host_already_in_inventory(hostname_or_id)
+  def host_already_in_inventory(id)
     response = Platform.connection.get(
       @url, {
-        hostname_or_id: hostname_or_id
+        id: id
```
Comment (akofink) [inline app/services/host_inventory_api.rb]: The inventory API does not accept an `id` param like this.

```
http  GET https://ci.cloud.redhat.com/api/inventory/v1/hosts id=='38537355-a8e3-4290-9b59-2205e503742d'
HTTP/1.1 400 BAD REQUEST
Transfer-Encoding: chunked
X-Powered-By: Express
connection: keep-alive
content-type: application/problem+json
date: Wed, 11 Mar 2020 17:24:54 GMT
server: openresty/1.13.6.1
set-cookie: bd2e829019cc5851647126f4f0834941=4fc6cf0ca78d2fa39bf6c07ccc15178f; path=/; HttpOnly; Secure
x-content-type-options: nosniff
x-fakamai-app-hint: API
x-fakamai-downstream-host: https://api-gateway-ci.5a9f.insights-dev.openshiftapps.com
x-fakamai-env: ci
x-rh-insights-request-id: 24ebf2dd25a145668eda4bfd2597239b

{
    "detail": "Extra query parameter(s) id not in spec",
    "status": 400,
    "title": null,
    "type": "about:blank"
}


http GET https://ci.cloud.redhat.com/api/inventory/v1/hosts/38537355-a8e3-4290-9b59-2205e503742d
HTTP/1.1 200 OK
Transfer-Encoding: chunked
X-Powered-By: Express
cache-control: private
connection: keep-alive
content-type: application/json
date: Wed, 11 Mar 2020 17:30:50 GMT
server: openresty/1.13.6.1
set-cookie: bd2e829019cc5851647126f4f0834941=546c250d4ef07df905cfb33195ad89ad; path=/; HttpOnly; Secure
x-content-type-options: nosniff
x-fakamai-app-hint: API
x-fakamai-downstream-host: https://api-gateway-ci.5a9f.insights-dev.openshiftapps.com
x-fakamai-env: ci
x-rh-insights-request-id: 2bd71ff758fd439b85e7d53bc70bc152

{
    "count": 1,
    "page": 1,
    "per_page": 50,
    "results": [
        {
            ...
        }
    ],
    "total": 1
}
```

File: app/services/xccdf_report_parser.rb [hunk 6]
```diff
@@ -13,7 +13,7 @@ class XccdfReportParser
   attr_reader :report_path, :test_result_file
 
   def initialize(report_contents, message)
-    raise ::EmptyMetadataError if message['metadata'].blank?
+    raise ::MessageFormatError unless valid_message_format?(message)
```
Comment (akofink) [inline app/services/xccdf_report_parser.rb]: We also need to raise an error if the host is not found in the inventory.

File: test/services/host_inventory_api_test.rb [hunk 7]
```diff
@@ -30,25 +30,8 @@ class HostInventoryApiTest < ActiveSupport::TestCase
...
-    assert_equal @host.id, @api.create_host_in_inventory['id']
-  end
-
   test 'inventory_host for host already in inventory' do
     @api.expects(:host_already_in_inventory).returns(@host)
-    @api.expects(:create_host_in_inventory).never
-    assert_equal @host, @api.inventory_host
-  end
-
-  test 'inventory_host for host not already in inventory' do
```
Comment (akofink) [inline test/services/host_inventory_api_test.rb]: We still need this test, but updated to assert that nil is returned when the host isn't found.

File: test/services/host_inventory_api_test.rb [hunk 8]
```diff
@@ -30,25 +30,8 @@ class HostInventoryApiTest < ActiveSupport::TestCase
...
-    )
-
-    @connection.expects(:post).returns(response)
-    assert_equal @host.id, @api.create_host_in_inventory['id']
-  end
-
   test 'inventory_host for host already in inventory' do
     @api.expects(:host_already_in_inventory).returns(@host)
-    @api.expects(:create_host_in_inventory).never
-    assert_equal @host, @api.inventory_host
```
Comment (akofink) [inline test/services/host_inventory_api_test.rb]: This assertion should still hold.

File: test/services/reports_tar_reader_test.rb [hunk 9]
```diff
@@ -11,7 +11,7 @@ class ReportsTarReaderTest < ActiveSupport::TestCase
         XccdfReportParser.new(
           report,
           'account' => accounts(:test).account_number,
-          'id' => @host_id,
+          'id' => ::UUID.generate,
```
Comment (akofink) [inline test/services/reports_tar_reader_test.rb]: I don't understand why this change is necessary - I assume some test failed, but I don't understand why.

File: test/services/xccdf_report_parser_test.rb [hunk 10]
```diff
@@ -184,6 +184,29 @@ class TestParser < ::XccdfReportParser
...
+        TestParser.new(
+          'fakereport',
+          'account' => accounts(:test).account_number,
+          'b64_identity' => 'b64_fake_identity',
+          'metadata' => { 'fqdn': '123' }
+        )
+      end
+    end
+
+    should 'raise error if metadata fqdn is not present' do
```
Comment (akofink) [inline test/services/xccdf_report_parser_test.rb]: Why is fqdn required? It seems like we don't use it at all - we update our host's name from the fqdn returned by the inventory request.

File: test/services/reports_tar_reader_test.rb (hunk 9 above)
Comment (dLobatog) [reply test/services/reports_tar_reader_test.rb]: `@host_id` was nil here - so since `message['id']`was nil, it failed

File: test/services/host_inventory_api_test.rb (hunk 8 above)
Comment (dLobatog) [reply test/services/host_inventory_api_test.rb]: This assertion was not removed, see line 52

File: test/services/xccdf_report_parser_test.rb (hunk 10 above)
Comment (dLobatog) [reply test/services/xccdf_report_parser_test.rb]: Before I removed create_host_in_inventory, it was used - now it can be removed and we can validate just `MissingId`

File: app/services/xccdf_report_parser.rb [hunk 11]
```diff
@@ -1,7 +1,7 @@
 # frozen_string_literal: true
 
-# Error to raise if no metadata is available
-class EmptyMetadataError < StandardError; end
+# Error to raise if id is not available
+class MissingIdError < StandardError; end
```
Comment (dLobatog) [inline app/services/xccdf_report_parser.rb]: It doesn't look like we need metadata anymore - we just need to make sure the ID is there - so that's what I'm checking for now.

File: app/services/xccdf_report_parser.rb [hunk 12]
```diff
@@ -30,6 +29,10 @@ def check_report_format
     )
   end
 
+  def valid_message_format?(message)
```
Comment (dLobatog) [inline app/services/xccdf_report_parser.rb]: It might seem odd to have a separate method for this but I think it's likely we'll have to validate other fields in the message in the future - so just making this method a bit bigger would work for that

File: test/services/host_inventory_api_test.rb [hunk 13]
```diff
@@ -30,26 +30,16 @@ class HostInventoryApiTest < ActiveSupport::TestCase
...
     @api.expects(:host_already_in_inventory).returns(@host)
-    @api.expects(:create_host_in_inventory).never
     assert_equal @host, @api.inventory_host
   end
 
   test 'inventory_host for host not already in inventory' do
-    @api.expects(:host_already_in_inventory).twice
-    @api.expects(:create_host_in_inventory).returns(@host)
-    assert_equal @host, @api.inventory_host
+    assert_raises(::InventoryHostNotFound) do
```
Comment (dLobatog) [inline test/services/host_inventory_api_test.rb]: We do want to raise an error in this case - no hosts should be created by us, so if the message does not contain the host ID of a host in the inventory, let's error out.

File: app/jobs/parse_report_job.rb [hunk 14]
```diff
@@ -28,7 +28,7 @@ def save_all
     notify_payload_tracker(:processing, "Job #{jid} is now processing")
     parser.save_all
     notify_payload_tracker(:success, "Job #{jid} has completed successfully")
-  rescue ::EmptyMetadataError, ::WrongFormatError => e
+  rescue ::MissingIdError, ::WrongFormatError, ::InventoryHostNotFound => e
```
Comment (dLobatog) [inline app/jobs/parse_report_job.rb]: This merely ensures the expected errors (missing ID, or missing host in the inventory) are logged in payload tracker and in our logs

File: app/services/concerns/xccdf/hosts.rb (hunk 4 above)
Comment (dLobatog) [reply app/services/concerns/xccdf/hosts.rb]: Said and done

File: app/services/host_inventory_api.rb [hunk 15]
```diff
@@ -6,59 +6,34 @@
...
   end
 
-  def host_already_in_inventory(hostname_or_id)
+  def host_already_in_inventory(id)
     response = Platform.connection.get(
-      @url, {
-        hostname_or_id: hostname_or_id
-      },
-      X_RH_IDENTITY: @b64_identity
+      "#{@url}/#{id}", {}, X_RH_IDENTITY: @b64_identity
```
Comment (dLobatog) [inline app/services/host_inventory_api.rb]: This check is much simpler than before - I could remove the argument and use the instance variable but it's handy to be able to reuse the same `HostInventoryAPI` object to check for any inventory ID :smile:
